map/framework.cpp

```diff
@@ -2166,18 +2172,30 @@ std::optional<place_page::Info> Framework::BuildPlacePageInfo(
     }
   }
 
+  bool isBuildingSelected = false;
   if (isFeatureMatchingEnabled && !selectedFeature.IsValid())
+  {
     selectedFeature = FindBuildingAtPoint(buildInfo.m_mercator);
+    isBuildingSelected = selectedFeature.IsValid();
+  }
 
   bool showMapSelection = false;
   if (selectedFeature.IsValid())
   {
+    // Selection circle should match feature
     FillFeatureInfo(selectedFeature, outInfo);
-    if (buildInfo.m_isLongTap)
-      outInfo.SetMercator(buildInfo.m_mercator);
+
+    if (isBuildingSelected)
+    {
+      // Calculate distance between building centroid and tap point
+      double distancePx = (m_currentModelView.GtoP(buildInfo.m_mercator) - m_currentModelView.GtoP(outInfo.GetMercator())).Length();
+      // If user taps on a building far from it's centroid (>10px) move selection circle to tap position
+      if (distancePx > kBuildingCentroidThreshold * df::VisualParams::Instance().GetVisualScale())
+        outInfo.SetMercator(buildInfo.m_mercator);
+    }
```

Comment:
```suggestion
    if (isBuildingSelected)
      outInfo.SetMercator(buildInfo.m_mercator);
```

If a building has something visual at its centroid (a POI, a housenumber..) then the selection will be snapped to it already via the code before this one.
If there is nothing visual at the centroid, then why snap to it at all? Leave the selection cirlce at the tapped point.
WDYT?

(I've tested this change and it works fine for me)

(and when https://github.com/organicmaps/organicmaps/issues/4998 is fixed properly then we'll be able to remove all "building"-specific logic from here, incl. FindBuildingAtPoint())

Reply:
Also your current snapping code doesn't work at z14-15 anyway (see https://github.com/organicmaps/organicmaps/pull/7876#issuecomment-2117967830), so my change will make it more consistent across zoom levels, actually.

Reply:
Accepted your solution. It works well